### PR TITLE
Fix: Samples don't reflect downstream entities from stitched subgraphs (#3095)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
https://github.com/DataBiosphere/azul/issues/3095

Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by noah-aviel-dove** [comment]
Repeat Hannes's patching of the unit tests (https://github.com/DataBiosphere/azul/issues/2868#issuecomment-792501645) and check the actual size of the contributions for the bundle that he analyzed.

**Comment by noah-aviel-dove** [comment]
I repeated [@hannes-ucsc 's analysis of contribution size] from the issue we're hoping to avoid re-invoking, and the results look good. The largest contribution for the bundle is only 321 KB.

```
$ git show --name-only --oneline
c6adb852 (HEAD -> issues/noah-aviel-dove/3095-missing-samples, github/issues/noah-aviel-dove/3095-missing-samples) Emit contributions for stitched samples (#3095)
src/azul/plugins/metadata/hca/transform.py
$ git diff
diff --git a/test/indexer/test_projects.py b/test/indexer/test_projects.py
index abda96f1..e4ef1c4b 100644
--- a/test/indexer/test_projects.py
+++ b/test/indexer/test_projects.py
@@ -105,6 +105,14 @@ class TestDataExtractorTestCase(IndexerTestCase):
                     self.assertEqual(num_files, len(file_ids))
                     self.assertEqual(num_files, len(set(file_ids)))
 
+    def test_debug_large_bundles(self):
+        import azul.indexer
+        from azul.plugins.repository.local import LocalSourceRef
+        spec = azul.indexer.SimpleSourceSpec.parse('data/:')
+        self._index_canned_bundle(azul.indexer.SourcedBundleFQID(source=LocalSourceRef(id=spec.name, spec=spec),
+                                                                 uuid='9697e762-96f2-5949-af7e-12fb4c4f22d0',
+                                                                 version='2021-02-08T10:30:00.000000Z'))
+
 
 if __name__ == "__main__":
     unittest.main()
```

With a breakpoint in `azul.indexer.index_service.IndexService.Index`:

Screenshot from 2021-11-03 04-05-09

**Review by danielsotirhos — CHANGES_REQUESTED** [review]
With this set in my environment.py
```
        'AZUL_CATALOGS': json.dumps({
            'it0largebundles': dict(atlas='hca',
                                    internal=True,
                                    plugins=dict(metadata=dict(name='hca'),
                                                 repository=dict(name='local')),
                                    sources=[
                                        'data/:'
                                    ]),
        }),
```
my run of IT failed with
```
======================================================================
ERROR: test_indexing (integration_test.IndexingIntegrationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/daniel/repo/azul3/test/azul_test_case.py", line 154, in wrapped
    return method(*args, **kwargs)
  File "/Users/daniel/repo/azul3/test/integration_test.py", line 340, in test_indexing
    _wait_for_indexer()
  File "/Users/daniel/repo/azul3/test/integration_test.py", line 319, in _wait_for_indexer
    self.azul_client.wait_for_indexer()
  File "/Users/daniel/repo/azul3/src/azul/azulclient.py", line 481, in wait_for_indexer
    self.queues.wait_to_stabilize()
  File "/Users/daniel/repo/azul3/src/azul/queues.py", line 261, in wait_to_stabilize
    raise Exception('The queues have stalled', final_length)
Exception: ('The queues have stalled', 13)
```
[it-failure-full-log.txt](https://github.com/DataBiosphere/azul/files/7469852/it-failure-full-log.txt)

**Comment by noah-aviel-dove** [comment]
The stall doesn't happen consistently for me, although I did see it once. The queues do eventually empty for me without errors.

**Review by danielsotirhos — APPROVED** [review]
Approved.

File: src/azul/plugins/repository/local/__init__.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        return Path(__file__).parent
+
+    def list_bundles(self,
+                     source: LocalSourceRef,
+                     prefix: str
+                     ) -> List[SourcedBundleFQID[LocalSourceRef]]:
+        source_prefix = source.spec.prefix.common
+        validate_uuid_prefix(source_prefix + prefix)
+        directory = self.local_path / source.spec.name
+        files = directory.glob(f'{source_prefix}{prefix}*{self._manifest_ext}')
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/local/__init__.py:117]: These files would be included in the Chalice `deployment.zip` and would have to be unpacked every time AWS Lambda initializes a Lambda execution context (a Docker container). Please move this source file and the cans to the attic …

File: test/integration_test.py [hunk 2]
```diff
@@ -346,9 +346,10 @@ def _wait_for_indexer():
                                           bundle_fqids=catalog.bundle_fqids)
 
         for catalog in catalogs:
-            self._test_manifest(catalog.name)
-            self._test_dos_and_drs(catalog.name)
-            self._test_repository_files(catalog.name)
+            if config.catalogs[catalog.name].plugins['repository'].name != 'local':
+                self._test_manifest(catalog.name)
+                self._test_dos_and_drs(catalog.name)
+                self._test_repository_files(catalog.name)
```
Comment (hannes-ucsc) [inline test/integration_test.py:352]: … and remove this …

File: test/integration_test.py [hunk 3]
```diff
@@ -1244,9 +1245,9 @@ def file_name(part):
 
     def test_can_bundle_configured_catalogs(self):
         for catalog_name, catalog in config.catalogs.items():
-            if catalog.is_integration_test_catalog:
```
Comment (hannes-ucsc) [inline test/integration_test.py:1247]: ... and this, all in a separate commit so the history looks like this:

1) Add local plugin
2) Add cans
3) Move cans and local plugin to attic
4) Main commit

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Let's add a unit test for the smallest of the cans. The commit that adds the test should move the can out of the attic into the test data directory, it should precede the main commit, and pushed independently so I can see the assertion fail and then pass for the main commit.

File: test/indexer/test_hca_indexer.py [hunk 4]
```diff
@@ -1777,6 +1777,36 @@ def test_related_files_field_exclusion(self):
...
+                                  version='2020-12-10T10:30:00.000000Z')
+        self._index_canned_bundle(bundle)
+        hits = self._get_all_hits()
+        entity_ids = {hit['_source']['entity_id'] for hit in hits}
+        expected_entity_ids = {
+            # Root bundle
+            '79fa91b4-f1fc-534b-a935-b57342804a70',
+            'abe1a013-af7a-45ed-8c26-f3793c24a1f4',
+            'c60bf214-0340-53c0-b75b-af84a82808d6',
+            # Stitched downstream samples
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1791]:
```suggestion
            # Stitched upstream samples
```

File: test/indexer/test_hca_indexer.py [hunk 5]
```diff
@@ -1777,6 +1777,36 @@ def test_related_files_field_exclusion(self):
...
+            '5fbd4bb4-e761-4345-99b8-bd5e141a8b94',
+            '8c0ef0ae-0687-4f2a-8a27-d833b7032a26',
+            '8c1b5489-7bf1-40c0-ba60-5e6b079e108a',
+            'a39685fd-b34a-4db7-b973-381919edb96e',
+            'c98d14ec-8761-45f7-8e86-e63c188bd97f',
+            'ddd10ecc-f7b7-4370-82cf-14e30987b751',
+            'ee2bfb40-7aa4-4c78-b78a-91aa69c7dc6e',
+            'f37a0dec-b7ba-4b34-8e51-732fb5ac7213',
+            'f47bb6a8-a753-4d40-8254-28fee9d03c1b',
+            'f5d590c5-b307-4b99-af98-8ddba1e2a266'
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1806]: This only proves that something is contributed to the upstream samples, but not what. The test should assert the expectation from the issue description. And if possible I would like the assertion to be less literal.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/indexer/test_hca_indexer.py [hunk 6]
```diff
@@ -1778,34 +1778,31 @@ def test_related_files_field_exclusion(self):
...
+        def is_sample_aggregate(hit):
+            entity_type, aggregate = self._parse_index_name(hit)
+            return entity_type == 'samples' and aggregate
+
+        samples = [
+            hit['_source']['contents']
+            for hit in self._get_all_hits()
+            if is_sample_aggregate(hit)
+        ]
+        self.assertEqual(len(samples), 15)
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1797]: expectation should be first argument

File: test/indexer/test_hca_indexer.py [hunk 7]
```diff
@@ -1777,6 +1777,33 @@ def test_related_files_field_exclusion(self):
...
+
+        for sample in samples:
+            analysis_file_formats = [
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1800]:
```suggestion
            analysis_file_formats = {
```

File: test/indexer/test_hca_indexer.py [hunk 8]
```diff
@@ -1777,6 +1777,33 @@ def test_related_files_field_exclusion(self):
...
+                file['file_format']
+                for file in sample['files']
+                if 'DCP/2 Analysis' in file['file_source']
+            ]
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1804]:
```suggestion
            }
```

File: test/indexer/test_hca_indexer.py [hunk 9]
```diff
@@ -1777,6 +1777,33 @@ def test_related_files_field_exclusion(self):
...
+            self.assertIn('bam', analysis_file_formats)
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:1805]:
```suggestion
            self.assertEquals({'bam', 'loom', …} , analysis_file_formats)
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
And let's completely remove the other cans in the attic from the history so they don't take up any space in the repo. They're just a little too big compared to the value they add.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
The big cans need to be squashed out completely. As long as there is a commit reachable from the develop branch that has those large files (and there would be if we merged this as is), they will take up space in the repository. I wasn't clear about this so no bump. And I would like to see the commit that adds the test and the one that adds the fix to be pushed separately, one more time. Don't need fixups this time.

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
This does require reindexing to fix the index, so the last commit should carry the `[r]` tag, and the PR the `reindex` label.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
If we promote this today, we would have a genuine reason for reindexing.

**Pull request merged**